util/src/main/scala/scala/scalanative/util/SharedScopeProvider.scala

```diff
@@ -0,0 +1,41 @@
+package scala.scalanative.util
+import java.util.concurrent.atomic.AtomicReference
+import java.util.function.UnaryOperator
+
+/** Unsafe manually managed scope provider allowing to share them between multiple contexts */
+final class SharedScopeProvider[Key] {
+
+  /** Registers shared usage of scope and executes given function.
+   *  If there its the no other usages of shared scope it would be closed
+   *  @param usageId unique identifier of scope usage
+   */
+  def identifiedBy[T](usageId: Key)(fn: Scope => T): T = {
+    try fn(use(usageId))
+    finally tryClose(usageId)
+  }
+
+  private[this] val usages       = new AtomicReference[Set[Key]](Set.empty[Key])
+  private[this] val scope: Scope = Scope.unsafe
+
+  private def tryClose(usageId: Key): Unit = {
+    usages.getAndUpdate {
+      new UnaryOperator[Set[Key]] {
+        override def apply(s: Set[Key]): Set[Key] = {
+          val updated = s - usageId
+          if (updated.isEmpty) {
+            scope.close()
+          }
```

Comment:
We cannot put that inside the `UnaryOperator`. The documentation of `getAndUpdate` specifically says that

> The function should be side-effect-free, since it may be re-applied when attempted updates fail due to contention among threads.

Instead, I think we should use `updateAndGet`, and perform this test on the *result* of `updateAndGet` instead of inside the lambda.